immer/detail/hamts/bits.hpp

```diff
@@ -18,43 +18,91 @@ namespace immer {
 namespace detail {
 namespace hamts {
 
-using bits_t   = std::uint32_t;
-using bitmap_t = std::uint32_t;
-using count_t  = std::uint32_t;
-using shift_t  = std::uint32_t;
 using size_t   = std::size_t;
 using hash_t   = std::size_t;
 
-template <bits_t B, typename T=count_t>
+template <size_t B>
+struct types
+{
+    static_assert(B < 6, "B > 6 is not supported.");
+
+    using bits_t   = std::uint32_t;
+    using bitmap_t = std::uint32_t;
+    using count_t  = std::uint32_t;
+    using shift_t  = std::uint32_t;
+};
+
+template <>
+struct types<6>
+{
+    using bits_t   = std::uint64_t;
+    using bitmap_t = std::uint64_t;
+    using count_t  = std::uint64_t;
+    using shift_t  = std::uint64_t;
+};
+
+template <size_t B>
+using bits_t   = types<B>::bits_t;
+
+template <size_t B>
+using bitmap_t = types<B>::bitmap_t;
+
+template <size_t B>
+using count_t  = types<B>::count_t;
+
+template <size_t B>
+using shift_t  = types<B>::shift_t;
+
+template <size_t B, typename T=count_t<B>>
 constexpr T branches = T{1} << B;
 
-template <bits_t B, typename T=size_t>
+template <size_t B, typename T=size_t>
 constexpr T mask = branches<B, T> - 1;
 
-template <bits_t B, typename T=count_t>
+template <size_t B, typename T=count_t<B>>
 constexpr T max_depth = (sizeof(hash_t) * 8 + B - 1) / B;
 
-template <bits_t B, typename T=count_t>
-constexpr T max_shift = max_depth<B, count_t> * B;
+template <size_t B, typename T=count_t<B>>
+constexpr T max_shift = max_depth<B, count_t<B>> * B;
 
 #define IMMER_HAS_BUILTIN_POPCOUNT 1
 
-inline count_t popcount(bitmap_t x)
+template <size_t B>
+inline auto popcount_fallback(bitmap_t<B> x)
 {
-#if IMMER_HAS_BUILTIN_POPCOUNT
-#  if defined(_MSC_VER)
-    return  __popcnt(x);
-#  else
-    return __builtin_popcount(x);
-#  endif
-#else
     // More alternatives:
     // https://en.wikipedia.org/wiki/Hamming_weight
     // http://wm.ite.pl/articles/sse-popcount.html
     // http://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel
     x = x - ((x >> 1) & 0x55555555);
     x = (x & 0x33333333) + ((x >> 2) & 0x33333333);
     return ((x + (x >> 4) & 0xF0F0F0F) * 0x1010101) >> 24;
```

Comment:
Be careful, this fallback only works for 32 bits, not for any size of bitmap!

Reply:
Ah, I will adjust constants based on wiki page
https://en.wikipedia.org/wiki/Hamming_weight
```
int popcount64c(uint64_t x)
{
    x -= (x >> 1) & m1;             //put count of each 2 bits into those 2 bits
    x = (x & m2) + ((x >> 2) & m2); //put count of each 4 bits into those 4 bits 
    x = (x + (x >> 4)) & m4;        //put count of each 8 bits into those 8 bits 
    return (x * h01) >> 56;  //returns left 8 bits of x + (x<<8) + (x<<16) + (x<<24) + ... 
}
```